src/python/pants/engine/README.md

```diff
@@ -75,52 +75,72 @@ In cases where this search detects any ambiguity (generally because there are tw
 can provide the same product with the same number of parameters), rule graph compilation will fail with
 a useful error message.
 
-### Datatypes
+### The `@dataclasstype` Type Constructor
 
-In practical use, builtin types like `str` or `int` do not provide enough information to disambiguate
-between various types of data in `@rule` signatures, so declaring small `datatype` definitions to
-provide a unique and descriptive type is highly recommended:
+In practical use, builtin types like `str` or `int` do not provide enough information to
+disambiguate between various types of data in `@rule` signatures, so declaring small, unique classes
+to encapsulate different states can help to make `@rule`sets with complex control flow between rules
+more declarative and self-documenting.
+
+#### Requirements for an Engine `Param`
+
+To use an instance of some class `C` as a `Param` to the engine:
+1. instances of `C` must be immutable,
+2. `__hash__` and `__eq__` must be implemented such that when constructing two separate instances of `C` with the same argument list `...`, `C(...) == C(...)` and `hash(C(...)) == hash(C(...))`.
+  - This can be ignored for singleton `Param`s.
+
+#### Motivation for the `@dataclasstype` Convenience Wrapper
+
+The engine currently provides a wrapper for [Python 3 `@dataclass`es](https://docs.python.org/3/library/dataclasses.html) in [`pants.util.objects.dataclasstype`](../util/objects.py), which achieves all three of:
+1. a compact and high-performance representation via `@dataclass`,
+2. static type checking via [the `dataclasses` mypy plugin](https://github.com/python/mypy/blob/master/mypy/plugins/dataclasses.py),
+3. an API decoupled from the underlying representation of the object via [`pants.util.objects.DatatypeMixin`](../util/objects.py)!
+
+The typical usage of `@dataclass` and `@dataclasstype` is to compose them to wrap a typed class definition:
+
+*Note that the [3rdparty `dataclasses` library](https://github.com/ericvsmith/dataclasses) must be in your BUILD file and `import`ed in Python 3.6!*
 
 ```python
-class FormattedInt(datatype(['content'])): pass
+from dataclasses import dataclass
+
+from pants.util.objects import dataclasstype
+
+
+@dataclasstype
+@dataclass
```

Comment:
```suggestion
@dataclass(frozen=True)
```

Reply:
Oh, wow, thanks for pointing this out (especially since I talk about immutability being the first requirement for engine params!!!)!!! After looking through https://docs.python.org/3/library/dataclasses.html#frozen-instances, I was surprised to find this excerpt:
> There is a tiny performance penalty when using frozen=True: __init__() cannot use simple assignment to initialize fields, and must use object.__setattr__().

@stuhood @jsirois @illicitonion @benjyw @Eric-Arellano were you aware of this? I suspect that would still be significantly faster than the current bloated meta-programmed implementation. In response to learning this cool new fact, in this PR I'm going to:
1. add a check in `dataclasstype` to ensure its argument has set `frozen=True`
2. make all example `dataclasstype`s use `frozen=True`
3. leave a comment about the performance concern.

Reply:
Done in e3240bc!